File: challenge_1_page_186/uuid_generator_test.rb
require 'test/unit/testcase'
require 'set'

require_relative 'uuid_generator'

class UUIDGeneratorTest < Test::Unit::TestCase
  CONCURRENT_USERS = 10
  ITERATIONS = 100

  false && def test_it_works
    generator = create_generator

    1.upto(10) do |expected_id|
      assert_equal expected_id, generator.generate
    end

    generator.stop
  end

  def test_concurrency
    generator = create_generator

    ITERATIONS.times do |i|
      generated_ids = Set.new

      expected_number_of_generated_ids = CONCURRENT_USERS

      min_id = i * CONCURRENT_USERS + 1
      max_id = (i+1) * CONCURRENT_USERS 
      expected_id_range = min_id..max_id

      threads = CONCURRENT_USERS.times.map do
        Thread.new do
          id = generator.generate
          generated_ids.add(id)
          Thread.current[:output] = id
        end
      end

      threads.map(&:join)

      assert(
        generated_ids.size == CONCURRENT_USERS,
        "Expected number of uniquely generated ids to equal #{expected_number_of_generated_ids}, but it was #{generated_ids.size}"
      )

      threads.each do |t|
        actual = t[:output]

        assert(
          expected_id_range === actual,
          "Expeceted ID #{actual} to be in range #{expected_id_range}"
        )
      end
    end

    generator.stop
  end

  private

  def create_generator
    UUIDGenerator.new
  end
end

File: challenge_1_page_186/uuid_generator_test.exs
ExUnit.start()

defmodule UUIDGeneratorTest do
  use ExUnit.Case

  @concurrency 10
  @iterations 1000

  test "it works" do
    pid = UUIDGenerator.start_server()

    Enum.each(1..10, fn n ->
      assert UUIDGenerator.generate(pid) == n
    end)

    UUIDGenerator.stop_server(pid)
  end

  test "it performs well under concurrency" do
    pid = UUIDGenerator.start_server()

    Enum.each(1..@iterations, fn i ->
      expected_min_id = (i - 1) * @concurrency + 1
      expected_max_id = i * @concurrency
      expected_id_range = expected_min_id..expected_max_id

      tasks =
        Enum.map(1..@concurrency, fn _j ->
          Task.async(fn -> UUIDGenerator.generate(pid) end)
        end)

      generated_ids = Enum.map(tasks, &Task.await/1)

      Enum.each(generated_ids, fn id -> 
        assert Enum.member?(expected_id_range, id)
      end)

      unique_generated_ids = MapSet.new(generated_ids)

      assert MapSet.size(unique_generated_ids) == @concurrency
    end)


    UUIDGenerator.stop_server(pid)
  end
end